text/0000-structural-records.md

```diff
@@ -0,0 +1,1298 @@
+
+The order in which we've put `alpha`, `beta`, and `gamma` here does not matter.
+We could have also written:
+
+```rust
+type TheRecord = { beta: bool, alpha: u8, gamma: &'static str };
+```
+
+As long as the type is a *permutation* of each pair of field name and field type,
+the type is the same. This also means that we can write: 
+
+```rust
+let my_record: TheRecord = { alpha: 42u8, gamma: "Dancing Ferris", beta: true };
+```
+
+### Implemented traits
+
+With respect to trait implementations, because the type is structural,
+and because there may be an unbound number of fields that can all be arbitrary
+identifiers, there's no way to define implementations for the usual traits
+in the language itself.
+
+[tuples]: https://doc.rust-lang.org/std/primitive.tuple.html
+
+Instead, the compiler will automatically provide trait implementations for the
+standard traits that are implemented for [tuples]. These traits are: `Clone`,
+`Copy`, `PartialEq`, `Eq`, `PartialOrd`, `Ord`, `Debug`, `Default`, and `Hash`.
+Each of these traits will only be implemented if all the field types of a struct
+implements the trait.
+
+For all of the aforementioned standard traits, the semantics of the
+implementations are similar to that of `#[derive(Trait)]` for named-field structs.
+
++ For cloning `{ alpha: A, beta: B, gamma: C }` the logic is simply:
+
+  ```rust
+  {
+      alpha: self.alpha.clone(),
+      beta: self.beta.clone(),
+      gamma: self.gamma.clone(),
+  }
+  ```
+
++ For `Default`, you would get:
+
+  ```rust
+  {
+      alpha: Default::default(),
+      beta: Default::default(),
+      gamma: Default::default(),
+  }
+  ```
+
++ For `PartialEq`, each field is compared with same field in `other: Self`.
+
++ For `ParialOrd` and `Ord`, lexicographic ordering is used based on
+  the name of the fields and not the order given because structural records
+  don't respect the order in which the fields are put when constructing or
+  giving the type of the record.
+
++ For `Debug` the same lexicographic ordering for `Ord` is used.
+  As an example, when printing out `my_record` as with:
+
+  ```rust
+  let my_record = { beta: true, alpha: 42u8, gamma: "Dancing Ferris" };
+  println!("{:#?}", my_record);
+  ```
+  
+  the following would appear:
+  
+  ```rust
+  {
+      alpha: 42,
+      beta: true,
+      gamma: "Dancing Ferris"
+  }
+  ```
+
++ For `Hash`, the same ordering of the fields as before is used and then
+  `self.the_field.hash(state)` is called on each field in that order.
+  For example:
+
+  ```rust
+  self.alpha.hash(state);
+  self.beta.hash(state);
+  self.gamma.hash(state);
+  ```
+
+For auto traits (e.g. `Send`, `Sync`, `Unpin`), if all field types implement
+the auto trait, then the structural record does as well.
+For example, if `A` and `B` are `Send`, then so is `{ x: A, y: B }`.
+
+A structural record is `Sized` if all field types are.
+If the lexicographically last field of a structural record is `!Sized`,
```

Comment:
I feel like forcing a particular name by in unnecessarily restrictive.  Couldn't the unsized field just be reordered to be last?  What's the imagined use case even for allowing unsized fields or is it just for completeness?  Wouldn't it be more parsimonious to disallow them initially?

Reply:
I think disallowing unsized fields would be a shame, since they are allowed in tuples, although even in tuples they are awkward to use, since you must place them last, and if they are not, you will get an odd error about the field not being sized, when you try to **use** it.

Reply:
We kind of want to allow any tuple field to be unsized, FWIW.
The current limitation is *almost entirely* an implementation limitation.

Reply:
> I feel like forcing a particular name by in unnecessarily restrictive. Couldn't the unsized field just be reordered to be last?

I don't think so no; The type system needs to see these in lexicographic ordering in all respects so that type identity can be determined by sorting fields lexicographically. Otherwise I think the well-formedness requirements would become difficult to reason about. This is also the way `Hash`, `Debug`, and behavior wrt. coherence is meant to work.

> What's the imagined use case even for allowing unsized fields or is it just for completeness? Wouldn't it be more parsimonious to disallow them initially?

Purely for consistency with tuples. I want to achieve parity with them.

> We kind of want to allow any tuple field to be unsized, FWIW.
> The current limitation is _almost entirely_ an implementation limitation.

If and when we do that then we can also lift the same restrictions for structural records but I think it would be premature to do it only for structural records.

Reply:
@Nemo157 is correct here, the problem already exists with trait objects, that are e.g. `dyn X + Y + Z`, and they're equal no matter the order they were written in.

I'd say disallow unsized records until we can support all fields being unsized.

Reply:
> At its simplest change from what's specified it could be something like lexicographic ordering, except a singular unsized field is moved to the end.

I suppose; just to make sure... how would this handle polymorphic cases where we have a `T: ?Sized`...? It might be `Sized` at instantiation but we don't know that yet...